{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e415e4bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Knn predefinie classification :  [1]\n",
      "Knn predefinie classification :  [6.46333333]\n"
     ]
    }
   ],
   "source": [
    "#KNN classification \n",
    "X=[[0],[1], [2],[3]]\n",
    "\n",
    "y1=[0,0,1,1]\n",
    "\n",
    "#valeur de K pour knn et choix de la distance \n",
    "\n",
    "from sklearn.neighbors import KNeighborsClassifier \n",
    "\n",
    "neigh= KNeighborsClassifier(n_neighbors =3  , metric = \"euclidean\")\n",
    "neigh.fit(X,y1)\n",
    "print('Knn predefinie classification : ',neigh.predict([[2.1]]) )\n",
    "\n",
    "#KNN Regression \n",
    "from sklearn.neighbors import KNeighborsRegressor\n",
    "Knn_regression=KNeighborsRegressor(3)#k=3\n",
    "y2=[15.3 ,8.6 ,6.5 ,4.29 ]\n",
    "Knn_regression.fit(X,y2)\n",
    "\n",
    "\n",
    "print('Knn predefinie classification : ',Knn_regression.predict([[2.1]]) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d0360a7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "from  collections import Counter\n",
    "import math\n",
    "def mean(labels):  #lables=target\n",
    "    \"\"\"calcul de la moyenne des k etiquettes\"\"\"\n",
    "    return sum(labels)/len(labels)\n",
    "def mode(labels): #C'est la (ou les) valeur(s) du caractère dont l'effectif est le plus grand.\n",
    "    \"\"\"calcul du mode des k etiquettes\"\"\"\n",
    "    return Counter(labels).most_common(1)[0][0] #counter donne le nombre d'occurence\n",
    "print(mode([0,1,2,2]))\n",
    "def euclidean_distance(point1,point2):\n",
    "    \"\"\"Calcul de la distance euclidienne\"\"\"\n",
    "    sum_squared_distance=0\n",
    "    for i in range(len(point1)):\n",
    "        sum_squared_distance+=math.pow(point1[i]-point2[i],2)\n",
    "    return math.sqrt(sum_squared_distance)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a746b711",
   "metadata": {},
   "outputs": [],
   "source": [
    "def knn(data, query, k,distance_utilisee, choix_knn_regress_ou_classif): #query :valeur a predire\n",
    "    \"\"\"Algorithme des k plus proches voisins knn\"\"\"\n",
    "    distances_et_positions = []\n",
    "    for index, example in enumerate(data):\n",
    "        distance=distance_utilisee(example[:-1],query) #chercher les voisins\n",
    "        distances_et_positions.append((distance, index))\n",
    "        #ici une liste de tuples sera triée selon les distances\n",
    "        sorted_distances_et_positions=sorted(distances_et_positions) #on fait le tri\n",
    "        k_proches_distances_and_indices=sorted_distances_et_positions[:k]\n",
    "        k_proches_classes=[]\n",
    "        for dist_indice in k_proches_distances_and_indices:\n",
    "            num=dist_indice[1];k_proches_classes+=[data[num][1]]\n",
    "            #7. si regression(choix_knn_regress_ou_classif = mean),\n",
    "            #retourner la moyenne des k etiquettes\n",
    "            #8. si classification, retourner le mode des k etiquettes\n",
    "            return k_proches_distances_and_indices,\\\n",
    "        choix_knn_regress_ou_classif(k_proches_classes),k_proches_classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "dd53032e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "knn implenté pour regression:\n",
      "[(6.659999999999997, 0)] \n",
      " 112.99 \n",
      " [112.99]\n",
      "knn predefinie-regression:\n",
      " [128.24666667]\n"
     ]
    }
   ],
   "source": [
    "from numpy import *\n",
    "taille_poids=array([[167.66,112.99],[182.38,136.49],\n",
    "                    [176.97,153.03],[173.96,142.34],[172.82,144.3],\n",
    "                    [175.19,123.3] ,[177.99,141.49],[178.53,136.46],\n",
    "                    [173.15, 112.37],[169.55, 127.45]])\n",
    "taille_en_cm=[161] #la valeur surlaquel on va appliquer le knn\n",
    "K=3;distance_utilisee=euclidean_distance;choix_knn=mean #regression\n",
    "R=knn(taille_poids,taille_en_cm,k,distance_utilisee,choix_knn)\n",
    "knn_dist_et_indices,prediction_de_poids,knn_lables=R\n",
    "print('knn implenté pour regression:')\n",
    "print(knn_dist_et_indices,'\\n',prediction_de_poids,'\\n',knn_lables)\n",
    "from sklearn import neighbors\n",
    "knn_regression =neighbors.KNeighborsRegressor(k)\n",
    "X=[[e] for e in taille_poids[:,0]];Y=[e for e in taille_poids[:,1]]\n",
    "knn_regression.fit(X,Y)\n",
    "print('knn predefinie-regression:\\n',knn_regression.predict([taille_en_cm]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f3018aa6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "résultat knn implémenté pour la classification :  \n",
      "[(11.0, 0)] \n",
      " 1 \n",
      " [1]\n",
      "[0]\n"
     ]
    }
   ],
   "source": [
    "#Données de Classification , Colonne 0 : age Colonne 1 : aime jeux ? \n",
    "\n",
    "age_aimer_jeux_electroniques=array([[22,1] , [23,1] , [21,1] , [18,1] , [19, 1] , [25,0] , [27, 0] , [29,0] , [31,0 ] , [45,0]])\n",
    "\n",
    "#age 33 ans aime ans aime t elle les jeux electronique \n",
    "\n",
    "age=[33] ; k=5 ; distance_utilisee= euclidean_distance ; choix_knn=mode\n",
    "R=knn(age_aimer_jeux_electroniques , age , k , distance_utilisee ,  choix_knn  )  \n",
    "\n",
    "knn_dist_et_indice , prediction_de_aimer_jeux , knn_labels =R \n",
    "\n",
    "print('résultat knn implémenté pour la classification :  ')\n",
    "print(knn_dist_et_indice , '\\n' , prediction_de_aimer_jeux , '\\n' , knn_labels)\n",
    "from sklearn.neighbors import KNeighborsClassifier \n",
    "\n",
    "neigh= KNeighborsClassifier(n_neighbors =3  , metric = \"euclidean\")\n",
    "\n",
    "X=[[e] for e in age_aimer_jeux_electroniques[:,0] ] #input \n",
    "\n",
    "Y= [e for e in age_aimer_jeux_electroniques[:,1] ]\n",
    "neigh.fit(X,Y)\n",
    "print(neigh.predict([age]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "37b0a9d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "score sur test-set par k=5 --> 85.0%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 598, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_classification.py\", line 179, in fit\n",
      "    return self._fit(X, y)\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 399, in _fit\n",
      "    self._check_algorithm_metric()\n",
      "  File \"C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 342, in _check_algorithm_metric\n",
      "    raise ValueError(\"Metric '%s' not valid. Use \"\n",
      "ValueError: Metric 'manhatten' not valid. Use sorted(sklearn.neighbors.VALID_METRICS['brute']) to get valid options. Metric can also be a callable function.\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\T460s\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_search.py:922: UserWarning: One or more of the test scores are non-finite: [0.82251978 0.79250092 0.8287479  0.82249632 0.83375574 0.81249941\n",
      " 0.80001502        nan        nan        nan        nan        nan\n",
      "        nan        nan]\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best parametrs: {'metric': 'euclidean', 'n_neighbors': 5}\n",
      "score sur test-set-parameteres automatiques=88.0%\n"
     ]
    }
   ],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.model_selection import GridSearchCV,train_test_split\n",
    "import pandas\n",
    "import numpy as np\n",
    "data=pandas.read_csv(\"C:\\images_chiffres_codees_niveau_de_gris.csv\")\n",
    "data=np.array(data); x=data[ 0:1000,1:];y=data[0:1000,0];\n",
    "x_train,x_test,y_train,y_test=train_test_split(x,y, test_size=0.2,random_state=0)\n",
    "model=KNeighborsClassifier(n_neighbors=5, metric='manhattan')\n",
    "model.fit(x_train,y_train); score=model.score(x_test,y_test)\n",
    "print(\"score sur test-set par k=5 --> \"+str(score*100)+'%')\n",
    "distances =['euclidean','manhatten'];valeurs_de_k= np.arange(1,8)\n",
    "parametres_grid={\"n_neighbors\":valeurs_de_k, 'metric':distances}\n",
    "model=KNeighborsClassifier()\n",
    "grid=GridSearchCV(model,parametres_grid,cv=3,scoring='accuracy')\n",
    "grid.fit(x_train,y_train); print(\"best parametrs:\",grid.best_params_)\n",
    "model=grid.best_estimator_ ;model.fit(x_train,y_train)\n",
    "score=model.score(x_test,y_test)\n",
    "print(\"score sur test-set-parameteres automatiques=\"+str(score*100)+'%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "565c28a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9da9d083",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c49df246",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
